In [15]:
import pandas as pd

In [16]:
df_fusionné = pd.read_csv('df_fusionné.csv')

In [17]:
df_fusionné_binaire = df_fusionné.sample(frac=0.2)
df_fusionné_binaire.iloc[:, 0] = df_fusionné_binaire.iloc[:, 0].apply(lambda x: 0 if x == 'N' else 1)

In [18]:
# Count the occurrences of each class
counts = df_fusionné_binaire.iloc[:, 0].value_counts()

# Find the number of instances in the minority class
min_count = counts.min()

# Create balanced DataFrame through undersampling
df_balanced_under = pd.concat([
    df_fusionné_binaire[df_fusionné_binaire.iloc[:, 0] == 0].sample(min_count),
    df_fusionné_binaire[df_fusionné_binaire.iloc[:, 0] == 1].sample(min_count)
])

# Shuffle the DataFrame to mix the classes
df_balanced_under = df_balanced_under.sample(frac=1)


In [19]:
# Count occurrences of 0 and 1 in the first column
value_counts = df_balanced_under.iloc[:, 0].value_counts()

# Display the counts
print(value_counts)

ColumnName
1    7625
0    7625
Name: count, dtype: int64


In [20]:
df_balanced_under

ColumnName         0         1         2         3         4         5  \
32521           1  0.564428  0.562515  0.567148  0.561858  0.562193  0.562140   
4393            1  0.490989  0.490951  0.490898  0.490839  0.490795  0.490750   
79268           1  0.405976  0.405980  0.405978  0.405972  0.405976  0.405980   
12191           1  0.490943  0.495637  0.479649  0.496694  0.489531  0.481669   
57643           1  0.497017  0.498338  0.502784  0.496685  0.498788  0.501286   
...           ...       ...       ...       ...       ...       ...       ...   
33583           1  0.468505  0.475315  0.475525  0.486948  0.489019  0.490550   
124761          1  0.512831  0.513078  0.513241  0.513332  0.513427  0.513486   
96233           1  0.510466  0.510441  0.510413  0.510379  0.510346  0.510311   
163327          0  0.507262  0.507727  0.508376  0.509076  0.509426  0.509738   
133753          0  0.505954  0.506736  0.506692  0.506668  0.508523  0.510521   

               6         7         8  ...       710       711       712  \
32521   0.557791  0.552267  0.550653  ...  0.514839  0.516868  0.515202   
4393    0.490705  0.490654  0.490565  ...  0.490887  0.490890  0.490908   
79268   0.405985  0.405986  0.405962  ...  0.405769  0.405789  0.405816   
12191   0.487432  0.496709  0.494757  ...  0.639220  0.638547  0.646542   
57643   0.499231  0.496043  0.496593  ...  0.353815  0.379979  0.392532   
...          ...       ...       ...  ...       ...       ...       ...   
33583   0.496394  0.503453  0.508047  ...  0.523470  0.519765  0.520428   
124761  0.513497  0.513463  0.513342  ...  0.506032  0.506646  0.507429   
96233   0.510273  0.510233  0.510195  ...  0.510432  0.510430  0.510432   
163327  0.509956  0.510235  0.511254  ...  0.506361  0.506692  0.506773   
133753  0.514028  0.517169  0.514428  ...  0.495763  0.497203  0.498395   

             713       714       715       716       717       718       719  
32521   0.512972  0.510599  0.508766  0.513420  0.523542  0.523411  0.513183  
4393    0.490930  0.490947  0.490963  0.490980  0.490992  0.491006  0.491012  
79268   0.405843  0.405868  0.405892  0.405914  0.405933  0.405952  0.405968  
12191   0.657139  0.656019  0.649270  0.659738  0.661253  0.652708  0.665375  
57643   0.400435  0.406813  0.412105  0.424603  0.441873  0.448265  0.443877  
...          ...       ...       ...       ...       ...       ...       ...  
33583   0.521991  0.524242  0.526227  0.521032  0.510181  0.510462  0.522599  
124761  0.508257  0.509039  0.509798  0.510527  0.511202  0.511866  0.512462  
96233   0.510435  0.510437  0.510439  0.510449  0.510458  0.510471  0.510480  
163327  0.506763  0.506771  0.506771  0.506784  0.506839  0.506845  0.506907  
133753  0.499484  0.500577  0.501630  0.502561  0.503414  0.504354  0.505231  

[15250 rows x 721 columns]

In [21]:
import pandas as pd
from sklearn.model_selection import train_test_split

X = df_balanced_under.iloc[:, 1:]  
y = df_balanced_under.iloc[:, 0] 

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [22]:
print(y_train.unique())

[1 0]


In [23]:
y_train = y_train.astype(int)
y_test = y_test.astype(int)

In [29]:
import tensorflow as tf
from keras.models import Sequential
from keras.layers import Conv1D, BatchNormalization, MaxPooling1D, Dropout, Flatten, Dense
from keras.optimizers import Adam

Model = Sequential()

Model.add(Conv1D(256, 10, strides=1, padding="same", activation="relu", input_shape=(720, 1)))
Model.add(BatchNormalization())
Model.add(MaxPooling1D(3, strides=2, padding="same"))

Model.add(Conv1D(256, 10, strides=1, padding="same", activation="relu"))
Model.add(Dropout(0.3))
Model.add(MaxPooling1D(3, strides=2, padding="same"))

Model.add(Conv1D(128, 10, strides=1, padding="same", activation="relu"))
Model.add(Dropout(0.3))
Model.add(MaxPooling1D(3, strides=2, padding="same"))

Model.add(Conv1D(64, 10, strides=1, padding="same", activation="relu"))
Model.add(Dropout(0.3))
Model.add(MaxPooling1D(3, strides=2, padding="same"))

Model.add(Conv1D(32, 10, strides=1, padding="same", activation="relu"))
Model.add(Dropout(0.3))
Model.add(MaxPooling1D(3, strides=2, padding="same"))

Model.add(Flatten())
Model.add(Dense(1024, activation='relu'))
Model.add(Dropout(0.3))

Model.add(Dense(256, activation='relu'))
Model.add(Dropout(0.3))

Model.add(Dense(64, activation='relu'))
Model.add(Dropout(0.3))

Model.add(Dense(units=1, activation='sigmoid'))  # Use 'sigmoid' for binary classification


In [30]:
print(Model.summary())

Model: "sequential_2"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv1d_10 (Conv1D)              │ (None, 720, 256)       │         2,816 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ batch_normalization_2           │ (None, 720, 256)       │         1,024 │
│ (BatchNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d_10 (MaxPooling1D) │ (None, 360, 256)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv1d_11 (Conv1D)              │ (None, 360, 256)       │       655,616 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_12 (Dropout)            │ (None, 360, 256)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d_11 (MaxPooling1D) │ (None, 180, 256)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv1d_12 (Conv1D)              │ (None, 180, 128)       │       327,808 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_13 (Dropout)            │ (None, 180, 128)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d_12 (MaxPooling1D) │ (None, 90, 128)        │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv1d_13 (Conv1D)              │ (None, 90, 64)         │        81,984 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_14 (Dropout)            │ (None, 90, 64)         │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d_13 (MaxPooling1D) │ (None, 45, 64)         │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv1d_14 (Conv1D)              │ (None, 45, 32)         │        20,512 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_15 (Dropout)            │ (None, 45, 32)         │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d_14 (MaxPooling1D) │ (None, 23, 32)         │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten_2 (Flatten)             │ (None, 736)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_6 (Dense)                 │ (None, 1024)           │       754,688 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_16 (Dropout)            │ (None, 1024)           │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_7 (Dense)                 │ (None, 256)            │       262,400 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_17 (Dropout)            │ (None, 256)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_8 (Dense)                 │ (None, 64)             │        16,448 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_18 (Dropout)            │ (None, 64)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_9 (Dense)                 │ (None, 1)              │            65 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 2,123,361 (8.10 MB)

 Trainable params: 2,122,849 (8.10 MB)

 Non-trainable params: 512 (2.00 KB)

None


In [31]:
# Define callbacks
Early_Stopper = tf.keras.callbacks.EarlyStopping(monitor="loss", patience=3, mode="min")


In [32]:
# Compile the model
Model.compile(optimizer=Adam(), loss='binary_crossentropy', metrics=['accuracy'])

In [33]:
Conv1D_Model = Model.fit(X_train, y_train, epochs=70, validation_data=(X_test, y_test), callbacks=[Early_Stopper])

Epoch 1/70
382/382 ━━━━━━━━━━━━━━━━━━━━ 24s 56ms/step - accuracy: 0.6795 - loss: 0.5878 - val_accuracy: 0.5584 - val_loss: 0.9485
Epoch 2/70
382/382 ━━━━━━━━━━━━━━━━━━━━ 21s 54ms/step - accuracy: 0.7902 - loss: 0.4553 - val_accuracy: 0.6318 - val_loss: 0.8058
Epoch 3/70
382/382 ━━━━━━━━━━━━━━━━━━━━ 21s 54ms/step - accuracy: 0.8173 - loss: 0.4216 - val_accuracy: 0.6174 - val_loss: 0.8395
Epoch 4/70
382/382 ━━━━━━━━━━━━━━━━━━━━ 21s 54ms/step - accuracy: 0.8372 - loss: 0.3964 - val_accuracy: 0.8023 - val_loss: 0.4309
Epoch 5/70
382/382 ━━━━━━━━━━━━━━━━━━━━ 21s 54ms/step - accuracy: 0.8370 - loss: 0.3757 - val_accuracy: 0.5695 - val_loss: 1.0270
Epoch 6/70
382/382 ━━━━━━━━━━━━━━━━━━━━ 21s 54ms/step - accuracy: 0.8616 - loss: 0.3384 - val_accuracy: 0.5538 - val_loss: 1.1014
Epoch 7/70
382/382 ━━━━━━━━━━━━━━━━━━━━ 21s 54ms/step - accuracy: 0.8636 - loss: 0.3372 - val_accuracy: 0.7590 - val_loss: 0.5056
Epoch 8/70
382/382 ━━━━━━━━━━━━━━━━━━━━ 21s 54ms/step - accuracy: 0.8703 - loss: 0.3143 - 

In [35]:
Model.save('Conv1D_Model.h5')  # Saves the model to a HDF5 file

In [36]:
Model_Results = Model.evaluate(X_test,y_test)
print("LOSS:  " + "%.4f" % Model_Results[0])
print("ACCURACY:  " + "%.4f" % Model_Results[1])

96/96 ━━━━━━━━━━━━━━━━━━━━ 1s 11ms/step - accuracy: 0.8812 - loss: 0.3155
LOSS:  0.3085
ACCURACY:  0.8816


In [37]:
# Assuming your model is named 'Model'
y_pred_probs = Model.predict(X_test)
y_pred = (y_pred_probs > 0.5).astype("int32")  # Convert probabilities to binary labels


96/96 ━━━━━━━━━━━━━━━━━━━━ 1s 10ms/step


In [38]:
from sklearn.metrics import confusion_matrix, precision_score, recall_score, f1_score, roc_auc_score, roc_curve, log_loss

# Confusion Matrix
conf_matrix = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:\n", conf_matrix)

# Precision
precision = precision_score(y_test, y_pred)
print(f"Precision: {precision:.2f}")

# Recall
recall = recall_score(y_test, y_pred)
print(f"Recall: {recall:.2f}")

# F1 Score
f1 = f1_score(y_test, y_pred)
print(f"F1 Score: {f1:.2f}")

# ROC-AUC Score
roc_auc = roc_auc_score(y_test, y_pred_probs)
print(f"ROC AUC Score: {roc_auc:.2f}")

# Now, you can calculate the log loss
logloss = log_loss(y_test, y_pred_probs)
print(f"Log Loss: {logloss:.2f}")

Confusion Matrix:
 [[1355  175]
 [ 186 1334]]
Precision: 0.88
Recall: 0.88
F1 Score: 0.88
ROC AUC Score: 0.95
Log Loss: 0.31


Confusion Matrix:
 [[1456  106]
 [ 640  834]]
Precision: 0.89
Recall: 0.57
F1 Score: 0.69
ROC AUC Score: 0.86
Log Loss: 0.65

Confusion Matrix:
 [[1357  127]
 [ 325 1187]]
Precision: 0.90
Recall: 0.79
F1 Score: 0.84
ROC AUC Score: 0.95
Log Loss: 0.33

Confusion Matrix:
 [[1401  104]
 [ 947  576]]
Precision: 0.85
Recall: 0.38
F1 Score: 0.52
ROC AUC Score: 0.81
Log Loss: 0.64
